framework: tlvf: fix wrong memory move on multiple allocations

morantr opened the pull request:
Problem:
When allocating/creating a new object in a dynamic list, the TLVF moving
all the memory from the beginning of the list to the end of the buffer
minus a new allocated size.
This could work only if increasing list size one time only when the
list size is 0.
When trying to do it multiple times, it caused memory override.

Solution:
Move the memory from (list beginning + its current size) instead of
(beginning of the list).

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
framework: tlvf: fix wrong memory move on multiple allocations

Problem:
When allocating/creating a new object in a dynamic list, the TLVF moving
all the memory from the beginning of the list to the end of the buffer
minus a new allocated size.
This could work only if increasing list size one time only when the
list size is 0.
When trying to do it multiple times, it caused memory override.

Solution:
Move the memory from (list beginning + its current size) instead of
(beginning of the list).

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

arnout commented on framework/tlvf/AutoGenerated/src/tlvf/ieee_1905_1/tlvWscM2.cpp:
```diff
@@ -454,7 +454,10 @@ bool tlvWscM2::add_encrypted_settings(std::shared_ptr<WSC::cWscAttrEncryptedSett
-    if (ptr->getStartBuffPtr() != (uint8_t *)m_encrypted_settings) {
+    uint8_t *src = (uint8_t*) m_encrypted_settings;
+    if (!m_parse__) {
+    }
```
This looks a bit weird, but OK, generated code...

tomereli replied:
Still, I guess the logic in tlvf.py can be changed so no empty logics such as this are created.

tomereli reviewed (changes requested):
add a test for it on top of the test in https://github.com/prplfoundation/prplMesh/pull/136

arnout commented on framework/tlvf/tlvf.py:
```diff
@@ -1006,7 +1006,16 @@ def addClassVarLenMethods(self, obj_meta, param_type, param_name, param_meta, pa
-            lines_cpp.append( "%sif (ptr->getStartBuffPtr() != (uint8_t *)m_%s) {" % (self.getIndentation(1), param_name) )
+            lines_cpp.append("%suint8_t *src = (uint8_t*) m_%s;" % (self.getIndentation(1), param_meta.name))
+            lines_cpp.append("%sif (!m_parse__) {" %self.getIndentation(1))
+            if param_length != None:
```
Since the body of the condition is only created of param_length != None, it would make sense to put the `if (!m_parse__)` also under that condition - i.e. move the condition one line higher (and include the closing brace too, of course).

tomereli commented on common/beerocks/tlvf/AutoGenerated/src/beerocks/tlvf/beerocks_message_bml.cpp:
```diff
@@ -181,8 +181,8 @@ bool cACTION_BML_NW_MAP_RESPONSE::alloc_buffer(size_t count) {
     m_lock_order_counter__ = 0;
-    uint8_t *src = (uint8_t *)m_buffer;
-    uint8_t *dst = (uint8_t *)m_buffer + len;
+    uint8_t *src = (uint8_t*) &m_buffer[*m_buffer_size];
```
remove space

tomereli commented on framework/tlvf/AutoGenerated/src/tlvf/ieee_1905_1/tlvDeviceBridgingCapability.cpp:
```diff
@@ -58,9 +58,12 @@ std::shared_ptr<cMacList> tlvDeviceBridgingCapability::create_bridging_tuples_li
+    uint8_t *src = (uint8_t*) m_bridging_tuples_list;
     if (!m_parse__) {
+        if (*m_bridging_tuples_list_length != 0) {
+            src = (uint8_t *)m_bridging_tuples_list_vector[*m_bridging_tuples_list_length - 1]->getBuffPtr();
```
Please add a comment explaining the logic in the tlvf.py so it will be clear.

tomereli commented on framework/tlvf/AutoGenerated/src/tlvf/ieee_1905_1/tlvWscM2.cpp:
```diff
@@ -454,7 +454,10 @@ bool tlvWscM2::add_encrypted_settings(std::shared_ptr<WSC::cWscAttrEncryptedSett
         return false;
     }
-    if (ptr->getStartBuffPtr() != (uint8_t *)m_encrypted_settings) {
+    uint8_t *src = (uint8_t*) m_encrypted_settings;
```
This doesn't look right...

LiorAmram replied:
why not?

tomereli replied:
because `m_encrypted_settings` is a pointer to the start of the list, but if this is not the first `create`-`add` then the src pointer should point to the last entry. I fixed this in #140

LiorAmram replied:
`src` points, at first, to `m_encrypted_settings`, that's the default value (in case that's the first create-add).
After that, the code checks if there are already created entries, and if so changes the value of `m_encrypted_settings`.
You don't see this code here because `m_encrypted_settings` is not a list but a single class, so there's no point to check the length of the list.
Look at line 245 in [tlvVarList.cpp](https://github.com/prplfoundation/prplMesh/pull/139/files/7e1c456e4daef585dea05e61209a637a5c0ce933#diff-63f6521e0ba10e43bad9fbf707ec4087) and you'll understand what I mean.

tomereli commented on framework/tlvf/AutoGenerated/src/tlvf/ieee_1905_1/tlvWscM2.cpp:
```diff
@@ -454,7 +454,10 @@ bool tlvWscM2::add_encrypted_settings(std::shared_ptr<WSC::cWscAttrEncryptedSett
     }
-    if (ptr->getStartBuffPtr() != (uint8_t *)m_encrypted_settings) {
+    uint8_t *src = (uint8_t*) m_encrypted_settings;
+    if (!m_parse__) {
```
empty if - again, think this case got messed up, This is a simple list of uint8_t inside a class

LiorAmram replied:
The empty 'if' should be removed. Arnout already addressed the part in the python code that generates it.

tomereli commented on framework/tlvf/AutoGenerated/src/tlvf/test/tlvVarList.cpp:
```diff
@@ -105,9 +105,12 @@ std::shared_ptr<cInner> tlvTestVarList::create_complex_list() {
     m_lock_allocation__ = true;
-    uint8_t *src = (uint8_t *)m_complex_list;
-    uint8_t *dst = (uint8_t *)m_complex_list + len;
+    uint8_t *src = (uint8_t*) m_complex_list;
```
Here also - `src` is set twice

LiorAmram replied:
That's intended.
The first assignment is setting a default value (the beginning of the list), for the case of an empty list (in which the code won't enter the if and won't execute line 111).

tomereli commented on framework/tlvf/AutoGenerated/src/tlvf/test/tlvVarList.cpp:
```diff
@@ -125,7 +128,13 @@ bool tlvTestVarList::add_complex_list(std::shared_ptr<cInner> ptr) {
-    if (ptr->getStartBuffPtr() != (uint8_t *)m_complex_list) {
+    uint8_t *src = (uint8_t*) m_complex_list;
+    if (!m_parse__) {
+        if (m_complex_list_idx__ != 0) {
```
better change the if condition to `> 0` so it will be clear

tomereli commented on framework/tlvf/tlvf.py:
```diff
@@ -985,7 +985,7 @@ def addClassVarLenMethods(self, obj_meta, param_type, param_name, param_meta, pa
             lines_cpp.append( "%sm_%s__ = %s;" %(self.getIndentation(1), self.MEMBER_LOCK_ORDER_COUNTER, param_meta.list_index) )
             lines_cpp.append( "%sm_%s__ = true;" % (self.getIndentation(1), self.MEMBER_LOCK_ALLOCATION) )
-            lines_cpp.extend(self.addAllocationMarkers(obj_meta, "create", param_meta, param_length, True)) # Variable length lists support
+            lines_cpp.extend(self.addAllocationMarkersCreate(obj_meta, "create", param_meta, param_length, True)) # Variable length lists support
```
If you split to 3 separate functions - no need for the second argument. Why split in the first place?

morantr replied:
because for each one of: "create", "alloc", "add" has a different way to do it, and it is more convenient to find it int the tlvf.

tomereli commented on framework/tlvf/tlvf.py:
```diff
@@ -1085,13 +1094,34 @@ def addClassVarLenMethods(self, obj_meta, param_type, param_name, param_meta, pa
-    def addAllocationMarkers(self, obj_meta, func_name, param_meta, param_length, memmove):
+    def addAllocationMarkersCreate(self, obj_meta, func_name, param_meta, param_length, memmove):
+        lines_cpp = []
+        if param_meta.length_type == MetaData.LENGTH_TYPE_VAR:
```
You are missing one of my fixes. Should do the same for DYNAMIC length

tomereli replied:
```suggestion
        if param_meta.length_type == MetaData.LENGTH_TYPE_VAR or param_meta.length_type == MetaData.LENGTH_TYPE_DYNAMIC:
```

arnout commented on framework/tlvf/tlvf.py:
```diff
@@ -1085,13 +1094,34 @@ def addClassVarLenMethods(self, obj_meta, param_type, param_name, param_meta, pa
     # action is needed for these lists.
     ##########################################################################
-    def addAllocationMarkers(self, obj_meta, func_name, param_meta, param_length, memmove):
+    def addAllocationMarkersCreate(self, obj_meta, func_name, param_meta, param_length, memmove):
```
The Create and Alloc functions seem to be very similar. Isn't it possible to keep it as one function with some extra argument or something?

LiorAmram replied:
I think it's better to leave them as 2 separated methods, since they dictate different flows:
'create' requires a call to 'add' method in order to finish the process, while 'alloc' doesn't.
I believe different methods make it easier to follow this scheme.

tomereli replied:
Fine - so remove the func_name argument

morantr replied:
@LiorAmram can I remove it? This argument is func_name and it is being used inside the fucntion.
```
marker = "%s_%s_%s_%s" %(self.CODE_CLASS_ALLOC_INSERT, obj_meta.name, func_name, param_meta.name)
obj_meta.alloc_list.append((marker, func_name))
```

LiorAmram replied:
You no longer need this argument, since each method has its own generating function.
So inside `addAllocationMarkersCreate`, you know that `func_name=create`.

arnout commented on framework/tlvf/tlvf.py:
```diff
@@ -1006,7 +1006,16 @@ def addClassVarLenMethods(self, obj_meta, param_type, param_name, param_meta, pa
+            if param_length != None:
+                lines_cpp.append("%sif (m_%s_idx__ != 0) {" % (self.getIndentation(2), param_name))
+                ptr = "(uint8_t *)m_%s_vector[m_%s_idx__ - 1]->getBuffPtr()" %(param_meta.name,param_name)
+                lines_cpp.append("%ssrc = %s;" %(self.getIndentation(3), ptr))                    
```
If this wasn't generated code, I wouldn't even have a `m_%s_idx__` member and instead use:
```suggestion
                lines_cpp.append("%sif (m_%s_vector) {" % (self.getIndentation(2), param_name))
                lines_cpp.append("%ssrc = (uint8_t *)m_%s_vector.back()->getBuffPtr();" % self.getIndentation(3)
```
But maybe there are reasons to prefer to use `m_%s_idx__`

LiorAmram replied:
I agree, `m_%s_idx__` is mainly leftovers of old code (back in the days when we only supported simple lists), and now it's only needed for simple unknown length lists.
However, I think that changing the code to use `m_%s_idx__` only when necessary exceeds the scope of this PR and should be separated.

arnout reviewed (approved):
This is difficult to review, especially because those tlvf code generating functions are really hard to read and understand. It would have helped to split it up into individual patches for different aspects, with a better explanation of what you're doing. However, it looks OK at first sight, and I did look at the generated code and that definitely looks OK. So I'm approving. I do have a few comments but nothing essential, feel free to fix or to leave it as is.

I think this might be a fix of #137. Could you test that by applying #136 but without the "disable complex_list test" part, and check if this change fixes it? If yes, then ideally this should be included as part of #136. If not, please extend #136 with a test that exposes this issue, so you can prove it has been fixed.

tomereli commented on framework/tlvf/AutoGenerated/src/tlvf/ieee_1905_1/tlvWscM2.cpp:
```diff
@@ -454,7 +454,10 @@ bool tlvWscM2::add_encrypted_settings(std::shared_ptr<WSC::cWscAttrEncryptedSett
+    uint8_t *src = (uint8_t*) m_encrypted_settings;
+    if (!m_parse__) {
+    }
+    if (ptr->getStartBuffPtr() != src) {
```
This check is wrong - see #140 where I removed it

LiorAmram replied:
I don't understand why you removed it there...

tomereli replied:
I removed it there since on parse, The first create-add works, but then the second create add fails since the second entry's `ptr->getStartBuffPtr() != m_encrypted_settings` since the `m_encrypted_settings` points to the start of the list, so this check is only valid for the first entry.

tomereli replied:
Here is the same. Look at the tlvf test in #140 - create-add is called multiple times, which is how we add a list of classes of size > 1.
Then on parse, we call create-add until we get to the correct length - the second add fails on this check since `m_encrypted_settings != ptr->getStartBuffPtr`

```
        first  second
_______\ _ /__
|_______|__|__|
          /
        m_encrypted_settings

// second->getStartBuffPtr() != m_encrypted_settings
```

LiorAmram replied:
The change in this PR addresses this problem and fixes it.
Look at line 245 in [tlvVarList.cpp](https://github.com/prplfoundation/prplMesh/pull/139/files/7e1c456e4daef585dea05e61209a637a5c0ce933#diff-63f6521e0ba10e43bad9fbf707ec4087) and you'll see that this issue is resolved.

LiorAmram commented on framework/tlvf/tlvf.py:
```diff
@@ -1006,14 +1006,14 @@ def addClassVarLenMethods(self, obj_meta, param_type, param_name, param_meta, pa
 
-            lines_cpp.append("%suint8_t *src = (uint8_t*) m_%s;" % (self.getIndentation(1), param_meta.name))
-            lines_cpp.append("%sif (!m_parse__) {" %self.getIndentation(1))
+            lines_cpp.append("%suint8_t *src = (uint8_t *)m_%s;" % (self.getIndentation(1), param_meta.name))
```
Remove space before `*`
Should be `(uint8_t*)m_%s;`

morantr replied:
space should be between the type to the asterisks.

LiorAmram commented on framework/tlvf/tlvf.py:
```diff
@@ -1094,50 +1094,52 @@ def addClassVarLenMethods(self, obj_meta, param_type, param_name, param_meta, pa
         if param_meta.length_type == MetaData.LENGTH_TYPE_VAR:
             if memmove:
-                lines_cpp.append("%suint8_t *src = (uint8_t*) &m_%s[*m_%s];" % (self.getIndentation(1), param_meta.name, param_length))
+                lines_cpp.append("%suint8_t *src = (uint8_t *)&m_%s[*m_%s];" % (self.getIndentation(1), param_meta.name, param_length))
```
same

morantr replied:
space should be between the type to the asterisks.

LiorAmram commented on framework/tlvf/tlvf.py:
```diff
@@ -1094,50 +1094,52 @@ def addClassVarLenMethods(self, obj_meta, param_type, param_name, param_meta, pa
+        if param_meta.length_type == MetaData.LENGTH_TYPE_VAR or param_meta.length_type == MetaData.LENGTH_TYPE_DYNAMIC:
             if memmove:
-                lines_cpp.append("%suint8_t *src = (uint8_t*) m_%s;" % (self.getIndentation(1), param_meta.name))
+                lines_cpp.append("%suint8_t *src = (uint8_t *)m_%s;" % (self.getIndentation(1), param_meta.name))
```
same

LiorAmram commented on framework/tlvf/tlvf.py:
```diff
@@ -1094,50 +1094,52 @@ def addClassVarLenMethods(self, obj_meta, param_type, param_name, param_meta, pa
                 if param_length != None:
-                    lines_cpp.append("%sif (*m_%s != 0) {" % (self.getIndentation(2), param_length))
-                    ptr = "(uint8_t *)m_%s_vector[*m_%s - 1]->getBuffPtr()" %(param_meta.name,param_length)
+                    lines_cpp.append("%sif (m_%s_idx__ != 0) {" % (self.getIndentation(2), param_meta.name))
```
Actually here you can also use `>0` instead of `!=0`

arnout commented:
I did a quick check, and the parsing part of the test of #137 still fails. The generated bytes look OK though. Could you take a look at that?

Commit:
tlvf: Fix double swap for class members

When parsing complex list, calling add(create()) automatically swaps
the internal class, but then all of the outer
class including the inner class members
are again swapped at the end of the outer
class init() function, causing double swap
and wrong value for internal class member.

Fix this the same way we fixed  simple class member - by calling
class_swap() after add(), to revert the swap done by create(add()).

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: update autogenerated (fix double swap)

Update autogenerated files after fixing double swap for complex lists.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: organize test and add complex list test

Initial refactor for the test - take the current main into test_all() function,
add test_complex_list() which tests complex
lists as standalone, and add main()
which calls both.

Currently the complex list test passes (also was
used to find the double swap issue fixed in
the previous commits), but for some reason the full test still fails. This
will be fixed separately).

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on framework/tlvf/AutoGenerated/src/tlvf/test/tlvVarList.cpp:
```diff
@@ -125,7 +128,13 @@ bool tlvTestVarList::add_complex_list(std::shared_ptr<cInner> ptr) {
     }
-    if (ptr->getStartBuffPtr() != (uint8_t *)m_complex_list) {
+    uint8_t *src = (uint8_t *)m_complex_list;
+    if (!m_parse__) {
```
I'm not sure if this condition is correct. I think this may cause the error while parsing m_complex_list.

Indeed, while parsing, we also check if the pointer is the one we expected. For the first add, that is the m_complex_list pointer, but for subsequent adds, it should point to the end of the list, not the beginning.

arnout reviewed (approved):
I am not sure if the extra test function is really worth it - the other one should just work :-) But OK, fair enough.

I think I found the remaining problem, but since this fixes something, I'll already approve.
